neuron_explainer/activations/derived_scalars/README.md

```diff
@@ -1,63 +1,149 @@
+
+> [!NOTE]
+> In many places in the `neuron_explainer` codebase, **"dst"** is used in place of **"derived_scalar_type"**, or **"ds"** in place of **"derived_scalar"**, for conciseness.
+
+---
+
+## Core concepts
+
+### Key classes
+
+The key classes to understand are:
+
+#### `DerivedScalarType`
+
+- An enum containing names for all the derived scalars that have been defined.
+
+#### `PassType`
+
+- Forward or backward pass; derived scalars can depend on either forward pass activations, or backward pass gradients.
+
+#### `ActivationsAndMetadata`
+
+- A container for a set of pytorch tensors, corresponding to a certain `DerivedScalarType` and `PassType` at every layer index in the model for which it is defined.
+
+#### `ScalarDeriver`
+
+- An object specified by a certain `DerivedScalarType`, containing the necessary information to compute derived scalars for one or both `PassTypes`; specification often requires additional info in the `DstConfig`.
+
+#### `DstConfig`
+
+- A dataclass specifying any information required in constructing the `ScalarDeriver` beyond the `DerivedScalarType`. 
+
+  - For example, DSTs that use weight tensors must know what model they are being computed for, so that the correct weights can be accessed. 
+
+#### `ScalarSource`
+
+- An object specifying the inputs expected by a `ScalarDeriver`. This consists of either an `ActivationLocationType` and `PassType`, or a `ScalarDeriver` and `PassType` (for the case where a `ScalarDeriver`'s inputs themselves require a `ScalarDeriver` to compute) plus a `LayerIndexer`. These two types are referred to as `RawScalarSource` and `DerivedScalarSource` respectively. 
+
+  - **"Raw"** in this name refers to an activation is literally instantiated during a transformer forward/backward pass, and can be extracted using a hook at a particular line of code. 
+
+  - **"Derived"** refers to a quantity that can be computed from activations instantiated during a forward/backward pass (a superset of "Raw" activations).
+
+#### `LayerIndexer`
+
+- Defines a transformation to be applied to an `ActivationsAndMetadata` object such that each layer index of the output is the appropriate layer index for computing a given derived scalar. 
+
+  - For example, sometimes a derived scalar D at layer L (for L in `0...num_layers-1`) is computed by operating on activation A from layer L together with activation B from layer L0 (constant). In this case, we would apply a `ConstantLayerIndexer` to B, such that the activations of the `ActivationsAndMetadata` passed to the scalar deriver are from layer L0 of B, no matter the value of L.
+
+#### `RawActivationStore`
+
+- Contains `ActivationsAndMetadata` stored from a forward and possibly a backward pass, separated by `ActivationLocationType` and `PassType`.
+
+#### `DerivedScalarStore`: 
```

Comment:
Drop the trailing colon.

Reply:
Fixed this - [Fix derived_scalars readme](https://github.com/openai/transformer-debugger/pull/11/commits/c91b0fd47d813e328832f036f49544db06eb7fac)